### ISSUE TITLE
QEC fidelity for cx gates

### ISSUE THREAD

**Issue by quantumjim** [qiskit-advocate/qamp-spring-22#14]
### Description

The `cx` gate is crucial for quantum computing. However, any physical implementation will be imperfect in some way. Some measure of fidelity for any proposed or implemented `cx` is therefore an important benchmark.

In the era of fault-tolerant quantum computation, most `cx` gates applied on physical qubits will be dedicated to implementing quantum error correction (QEC). This therefore motivates the characterization of `cx` imperfects in the context of QEC.

In this project we will look at the effect of having a single imperfect `cx` within an otherwise perfect QEC protocol. The aim is to define a new QEC-inspired measure of fidelity that can be calculated from theoretically derived or experimentally measured tomography of a candidate `cx`.

### Deliverables

The aim is to define a new QEC-inspired measure of fidelity that can be calculated from theoretically derived or experimentally measured tomography of a candidate `cx`.

The results of this work should hopefully be included in a paper. In this case, it will hopefully be possible for the mentee to be an author.

### Mentors details

- Mentor
  - Name: James Wootton
  - GitHub ID: @quantumjim
  - What they do: QEC Researcher

### Number of mentees

1

### Type of mentees

- Mentee
  - Required:
    - Knowledge of linear algebra
  - Nice to have:
    - Knowledge of QEC

**Comment by jvscursulim** [comment]
I have interest in working on this project. My master's dissertation was about quantum error correction (unfortunately the text is only available in portuguese), in one part of the work I tried to reproduce the experiment described in https://arxiv.org/pdf/1912.09410.pdf using IBMQ Lagos. I also have a repository in my github where there is some QECC implemented on Qiskit (https://github.com/jvscursulim/Quantum-Error-Correcting-Codes).

**Comment by quantumjim** [comment]
Hi @jvscursulim and @abzsd.

I see you are both chose this as a possible project. Is that still the case?

**Comment by abzsd** [comment]
Yes, I am interested in being a part of this project.

**Comment by jvscursulim** [comment]
Yes, I still have interest in being part of this project.

**Comment by quantumjim** [comment]
I think we could have both of you on the project. Especially if we mainly communicate via Slack. Perhaps some time in the early afternoon (Swiss time) would suit everyone for any meetings all together.

**Comment by jvscursulim** [comment]
Looks good to me

**Comment by abzsd** [comment]
Yeah sure, great!!

**Comment by quantumjim** [comment]
@abzsd Are you on the Qiskit Slack? I can't seem to find you there.

**Comment by abzsd** [comment]
> @abzsd Are you on the Qiskit Slack? I can't seem to find you there.

Yeah, I'm there on Qiskit slack. My display name is Abhay K / Abhay_K.

**Comment by jvscursulim** [comment]
[#14 QEC fidelity for cx gates.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8441603/14.QEC.fidelity.for.cx.gates.pdf)
File for checkpoint 1

**Comment by abzsd** [comment]
Checkpoint 1 Presentation Slides :
[#14 QEC fidelity for cx gates.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8441711/14.QEC.fidelity.for.cx.gates.pdf)

**Comment by jvscursulim** [comment]
Since checkpoint 1 of the QAMP Spring 2022, I have been working on an analysis of the performance of different types of CNOT gates under some noise conditions: bit-flip error channel, depolarizing error channel, coherent errors and a combination of one error channel with coherent errors. The picture attached to this message, brings some results that I have obtained in the study of the CNOT constructed with  SWAP gate (In the picture this gate is represented by the two qubit gate called unitary). The main goal of this project is to build a benchmark to measure the quality of any type of CNOT gate. This kind of tool is important, once it can guide improvements in the development of CNOT gates. Achieving high fidelity CNOT gates is very important to quantum error correction, once these gates act as the entangling gates on the encoding phase of the protocol, but they also can be used to detect errors in the syndrome phase. Since we will need many CNOT gates, we need to guarantee that they have a low error rate, thus we can minimize the spreading of errors in the circuit and that we get a logical error as a result. 


In order to investigate the impact of a noisy CNOT gate in a quantum error correction circuit, we introduce one noisy gate in the syndrome part of the circuit to see how much impact this noisy gate has in the final result, I mean in the occurrence of a logical error. To do this task, I used the [[3,1,3]] repetition code with a qubit representing the environment that will induce a bit-flip in the code qubits. From the plots it is possible to see the following behavior: when CNOT gate infidelity increases we have more chances of having a occurence of a logical error. If the error occurs in the second qubit of the code, we have more chances of getting a logical error due the structure of the ancillae qubits. This analysis was also made for the standard CNOT, Hadamard-CZ CNOT and Molmer-Sorensen CNOT.

The next steps for the last month of the QAMP Spring 2022 will be: (1) An analysis of nonlinear behavior for some CNOTs as was pointed out in https://arxiv.org/abs/1808.03927; (2) An implementation and an analysis of the Floating CNOT gate and (3) The creation of a package to automatize the experiments done in my study.

![checkpoint2](https://user-images.githubusercontent.com/66651520/166827380-eeeda221-4ffb-4f76-86ea-928d99c529d7.jpg)

**Comment by abzsd** [comment]
I have been working on the finding the effect of combining the Randomized benchmarking technique with the Quantum Error correction approach, and trying to find how this particular arrangement causes the reduction in SPAM errors, or leads to increase in the cross-talk/correlated errors. The main goal of this project is to build a QEC-inspired benchmark to measure the fidelity of any CX gate. This technique would be able to direct the future developments in the CX gate, allowing them to try and achieve higher gate fidelities. The CX gates are very important as they are used as entangling gates in the encoding/decoding phase. Thus, the need for evaluation of the quality of a CX gate is very important. 

![RB_QEC_circ0](https://user-images.githubusercontent.com/54679619/166895576-a3245662-5ea0-4ff8-aa10-26aa9f563492.png)

The set of Clifford gates was applied in between the syndrome measurements and the final errors arising due to any noise in the circuit/system were studied. When using only the QEC technique the error was less(around 0.037 for '0' circuit and 0.024 for '1' circuit), but when the QEC technique was applied with the randomized benchmarking, then the error rates observed were much lesser(around 0.003 for '0' circuit and 0.002 for '1' circuit). Thus, indicating that maybe the syndrome measurements at the start and end of the circuit has affected/reduced the SPAM errors in the circuit, or maybe the reduction of cross-talk in some way.

![error_rates_QEC](https://user-images.githubusercontent.com/54679619/166896064-9b2a113e-db33-456c-a3ca-b44a5d59830e.png)
![error_rates_RB](https://user-images.githubusercontent.com/54679619/166896067-9188b5f1-81db-4f7a-99c8-8b1a8c8acfd1.png)

Further steps:
(1) Analysis of the constructed circuit and trying to find out how the Error per Clifford (EPC) varies
(2) An implementation of the various types of randomized benchmarking techniques and finding their effect on the error rates

**Comment by jvscursulim** [comment]
Checkpoint 3 (Final)

GitHub repositories:

https://github.com/jvscursulim/cnot-benchmark
https://github.com/jvscursulim/using_grover_algorithm_to_findout_the_distance_of_a_classical_error_correcting_code

[#14 Benchmarking noisy CX gates with QEC - Final showcase slides.pdf](https://github.com/qiskit-advocate/qamp-spring-22/files/8873205/14.Benchmarking.noisy.CX.gates.with.QEC.-.Final.showcase.slides.pdf)
